### ISSUE TITLE
Hmm ... 💭 

### ISSUE THREAD

**Issue by nelsonic** [dwyl/nextra-demo#4]
# Please Don't Read This! 🙏 

I don't _like_ being "negative" but I _know_ that I cannot fully convey my experience without _appearing_ "negative". 
Ultimately, Devs building `Nextra` have built it for their own needs and have included the features their community have requested/contributed. 

## Throwing Stones? 

I _live_ in a proverbial "[**_glass_ house**](https://www.britannica.com/dictionary/eb/qa/People-Who-Live-in-Glass-Houses-meaning)" with _extra large_ windows ... 🪟 
We have _many_ projects on `GitHub` with varying degrees of completion/maintenance. 
But none of the ones with _traction_ are abandoned/ignored. 
We regularly update the packages that have thousands of stars, millions of downloads and `people` using them
even when it takes time away from our _current_ work. 

# What?

At the time of wrIt is [Used by 13.3k](https://github.com/shuding/nextra/network/dependents) and has [**`10.9k Stars`**](https://github.com/shuding/nextra/stargazers) https://github.com/shuding/nextra

![image](https://github.com/user-attachments/assets/3d17cf72-572d-4c33-bb69-d3038804f055)

I should have voiced my "concerns" about this sooner, but I wanted to give it a "fair" chance `before` commenting.

_Specifically_ 
1. **337 issues open**. Most have **_zero_ comments** or ***acknowledgement*** from the Authors/Maintainers. 💬 0️⃣ 

https://github.com/shuding/nextra/issues?q=is%3Aissue+is%3Aopen+sort%3Acreated-desc
<img width="1150" alt="image" src="https://github.com/user-attachments/assets/0a4c8024-e071-45c4-ba2c-8170f754d905">

2. There are **67 PRs** with a few more than 2 years old and no comments from the Authors/Maintainers of the project.
https://github.com/shuding/nextra/pulls?q=is%3Apr+is%3Aopen+sort%3Acreated-asc
<img width="1150" alt="image" src="https://github.com/user-attachments/assets/a01dac76-4b09-4c7d-a39e-00f2e63a3bf3">

e.g: https://github.com/shuding/nextra/pull/315
![image](https://github.com/user-attachments/assets/46c46e79-d406-4552-98fa-753d57bb4679)

It's a ` +4 −5` and only  `2 Files changed` PR that _looks_ fine to me. 👌 
But no comment from the Authors/Maintainers ... ⏳ 
just a check-in-comment from @TheMikeyRoss **21 months later** indicating it was _still_ an issue for them. 😖 

> @maltejur saw your PR: https://github.com/shuding/nextra/pull/315 if you see this, are you still _using_ `Nextra`? 💭 

This is not to say that Pull Requests are _not_ being merged, they are.
e.g: https://github.com/shuding/nextra/pull/2981 merged by the 2<sup>nd</sup> highest contributor with no review. 🤷‍♂️ 

<img width="1007" alt="image" src="https://github.com/user-attachments/assets/dc39f33b-a78d-4a8d-962b-10cb9cb24d11">

Pay attention to the commit message: https://github.com/shuding/nextra/pull/2981/commits

Speaking of contributors: https://github.com/shuding/nextra/graphs/contributors

<img width="1351" alt="image" src="https://github.com/user-attachments/assets/9ee6709d-276c-4906-922b-92f813af53f3">

there are two main "guys", the owner - [Shu](https://github.com/shuding) who works at `Vercel` -  and [The Terminator](https://en.wikipedia.org/wiki/The_Terminator): https://github.com/dimaMachina 

<img width="1278" alt="image" src="https://github.com/user-attachments/assets/632318f2-fa38-455a-a031-1c3f21447274">

Who appears to be freelance: "hire me". 💭 
Both seem perfectly competent. As are the remaining frequent contributors. 👍  

# Build Failing on `main`

The build is failing on the `main` branch and there is no indication of priority for fixing it. ⏳ 
https://github.com/shuding/nextra/actions?query=branch%3Amain

<img width="1201" alt="image" src="https://github.com/user-attachments/assets/f90cfcf5-0e0a-42c1-9570-3429cfde794c">

With that many people _using_ your project, having a consistent `CI` should be a priority. 

# Homepage ...

At the time of writing this is the `Nextra` Home Page: https://nextra.site/

![image](https://github.com/user-attachments/assets/d3c6dfd7-c98d-4a13-b2bf-e80117ef5cb0)

It defaults to the `dark` theme because `system` is selected in the bottom-right of the screen. 
Dark cards on a dark background ... was this tested for contrast? 🤷 

## Page Speed Insights

https://pagespeed.web.dev/analysis/https-nextra-site/62p078my3z?form_factor=mobile
<img width="1103" alt="image" src="https://github.com/user-attachments/assets/3a9f22e8-1aa2-4735-8084-8cb8286f1505">

1.4s to load the page isn't bad. but they could definitely improve it if they wanted to.
![image](https://github.com/user-attachments/assets/a4aaf341-8197-44a3-a977-10c418e376fd)

Accessibility is a `90` (i.e. pass) despite not having a `lang` attribute:
![image](https://github.com/user-attachments/assets/ae364074-d02b-49d1-bd7e-65f292b4b72f)

You'd think a site _generator_ that offers `I18n` out-of-the-box would set the `lang` attribute ... 💭 
This brings us neatly to "features" ...

# What _Features_ does `Nextra` "give" us? 

`Google Analytics` _appears_ to work, but there is no documentation for it; ref: https://github.com/shuding/nextra/pull/2137 

## Sitemap + Indexing

Since the early web there has been a standard for indexing all pages on a site:  https://en.wikipedia.org/wiki/Sitemaps
There is a community package for this: https://www.npmjs.com/package/next-sitemap
But it has to be done _manually_ for your site: https://github.com/shuding/nextra/issues/419#issuecomment-1278407009

<img width="692" alt="image" src="https://github.com/user-attachments/assets/bb104da0-bec4-488d-bb5a-4eaf9fb0d1e2">

https://nextra.site/sitemap.xml
<img width="769" alt="image" src="https://github.com/user-attachments/assets/f71c92d0-7b49-4bd8-a9d6-dbbb4e9a0e37">

This is _basic_ SEO best practice, still 100% relevant in 2024: 
https://www.google.com/search?q=are+sitemaps+still+relevant

[https://developers.google.com/search/docs/crawling-indexing/robots](https://developers.google.com/search/docs/crawling-indexing/robots/create-robots-txt)

Similarly with `robots.txt` 
Visiting https://nextra.site/robots.txt you will see a `404`:
<img width="925" alt="image" src="https://github.com/user-attachments/assets/ad823f79-29cd-4253-99b8-ff140a9f57b2">
https://github.com/shuding/nextra/issues/1402 closed by [The Terminator] with no comment/explanation. 👎 
Referenced in: https://github.com/shuding/nextra/issues/1421 but doesn't _appear_ to be a feature in `V3` ...

Sadly, I don't have time to keep writing this right now. Gotta get back to work-work; which is _using_ `Nextra` for `/docs` ...
But I cannot help but feel that we are missing a few key features that are going to be _unnecessarily_ difficult to add to `Nextra` ... 💭

**Comment by TheMikeyRoss** [comment]
Something similar happened with https://github.com/jaredpalmer/tsdx
 
And seeing how shadcn-ui has almost 600 open PRs and is only maintained once a week (sometimes longer)

It seem that the story goes like this:
1. A competent and smart developer creates an open source project
2. The project gains popularity and becomes loved and used by many.
3. The dev behind the project gains recognition and reputation
4. Vercel hires that dev
6. Either, that open-source project gets less attention and slowly gets abandoned. Or, vercel acquires the project.


Similarly, Netlify killed content-layer. See https://github.com/contentlayerdev/contentlayer/issues/429

**Comment by nelsonic** [comment]
@TheMikeyRoss makes sense. 🙃 
`Vercel` are certainly following the `Blitzscaling` path: 🚀 
https://news.crunchbase.com/cloud/vercels-cloud-web-applications-funding-valuation-accel/
<img width="1312" alt="image" src="https://github.com/user-attachments/assets/84432cc2-48c6-4076-98ee-9c462adb1b0b">

At this sort of valuation they will have to sell to `Google` sooner or later. 🏦 
At that point all the _original_ creators of the various frameworks will cash-out 💰 
and sail off into the sunset ⛵  
leaving the projects in the hands of caretakers. 💭

**Comment by lacymorrow** [comment]
You say it like it's a bad thing! 
It's a shame when the project loses a maintainer, but even a small open-source project is a huge task for one person.

The missing piece is a community of maintainers.

> It seem that the story goes like this:
> 
> 1. A competent and smart developer creates an open source project
> 2. The project gains popularity and becomes loved and used by many.
> 3. The dev behind the project gains recognition and reputation
> 4. Vercel hires that dev
> 5. Either, that open-source project gets less attention and slowly gets abandoned. Or, vercel acquires the project.

**Comment by TheMikeyRoss** [comment]
> You say it like it's a bad thing!

The only bad thing is the project dying. And it's unfortunate that the project creator doesn't pass the torch to someone else or to a group of people.

Someone made [contentlayer2](https://github.com/timlrx/contentlayer2) just because the original creator of contentlayer didn't add maintainers despite many people offering their help.

**Comment by Zamoca42** [comment]
> Someone made [contentlayer2](https://github.com/timlrx/contentlayer2) just because the original creator of contentlayer didn't add maintainers despite many people offering their help.

Given this discussion about open-source project maintenance and the mention of contentlayer2, do you think a Nextra 2 might be needed for similar reasons?

**Comment by nelsonic** [comment]
@Zamoca42 no. That would be confusing given that `Nextra` is about to release a `v3` ...
The `Nextra` devs/maintainers are doing a _good_ job with the project - for the needs of the 👍 
What is required a _complete_ re-think of documentation sites. 💭

**Comment by lacymorrow** [comment]
What is the ideal future for this or other docs site? What would you like to see implemented?

**Comment by TheMikeyRoss** [comment]
When documentation sites are done right, they become a paid SaaS, See mintlify

**Comment by nelsonic** [comment]
@TheMikeyRoss indeed. [Mintlify](https://github.com/dwyl/product-ux-research/issues/50) does look good. At least _superficially_. 
There are still _many_ features missing e.g: [`Internationalization`](https://en.wikipedia.org/wiki/Internationalization_and_localization)

![image](https://github.com/user-attachments/assets/66b688ca-b733-4cec-93ce-1e8239cb1e75)

Surely I'm not the _first_ person who has searched their docs for `i18n` ... 🤷‍♂️ 
We have a _baseline_ (non-negotiable) need for our Docs site to be available in 3 languages.
So Mintlify immediately rules themselves out. 
Then there's the price: **`$4200/year`** for "Pro" is silly money for a docs site. 
Obvs in the context of the cost of a senior engineer it "pays for itself" very quickly if it has _all_ the needed features.
But it _really_ does not; not even _close_! And being **_Closed_ Source** means we have to _wait_ for the features to be added. ⏳ 
Pass. 

It's increasingly looking like we will need to craft our own custom solution. 🙃

**Comment by TheMikeyRoss** [comment]
> Then there's the price: $4200/year for "Pro" is silly money for a docs site.

Absolutely! And it feels extremely disproportional to what it actually costs them.

There's an oldie [bookstackapp.com](bookstackapp.com) (PHP 😞).
But I just dug into it and it seems they've been discussing i18n for 7 years now See [feature request](https://github.com/BookStackApp/BookStack/issues/569) 😅


I have a tingling feeling that there's a well-made, well-maintained, open-source documentation tool out there, I just haven't discovered it yet.

**Comment by Zamoca42** [comment]
I found two options worth looking into:
- [fumadocs](https://github.com/fuma-nama/fumadocs)
- [content-collections](https://github.com/sdorra/content-collections)

I haven't examined them in depth, but they seem promising as potential open-source documentation solutions. Might be worth checking out as alternatives to the pricier options discussed.

**Comment by LuchoTurtle** [comment]
@Zamoca42 thanks for sharing https://github.com/fuma-nama/fumadocs, I've been taking a look at it and it almost seems like a good "quasi drop-in replacement" for Nextra. It seems to have most (if not all) the same features as Nextra, with some other neat ones. It also looks to be more actively maintained 🤔

**Comment by TheMikeyRoss** [comment]
@Zamoca42 fumadocs was exactly what I was looking for, thanks!

**Comment by nelsonic** [comment]
Fuma doesn’t come _close_ to covering our needs. Here’s our list off the top of my head: https://github.com/dwyl/product-roadmap/issues/49

**Comment by nelsonic** [comment]
@TheMikeyRoss agreed, that's kind the point of opening this issue. 👍 
We have a decent list of features on our Roadmap 🗺️ 
and have surveyed the _existing_ landscape and found it lacking ... 😕 
`Nextra` doesn't have a "backend" by default/design. 🙅 
Which is _good_ for people who don't _want_ one, but makes building more advanced features tedious. 
We want to have `auth` so that we can protect certain content,
doing that in `Nextra` via [`next-auth`](https://github.com/dwyl/nextra-demo/pull/2/files#diff-84a5d130fb4cccb80f78601d140f1d5397dc0272f94cea672a8470539fcf6dc7) was _way_ over-complicated
and doesn't come _close_ to meeting our needs. 

Attempting to "_deeply integrate_" a **CMS** to a **docs site** is a **duct tape** solution. 🩹

[![image](https://github.com/user-attachments/assets/7e62aaf7-b266-41d6-951a-5fa1038a729a)](https://simple.wikipedia.org/wiki/Duct_tape#/media/File:Duct-tape_Moving_Van.jpg)

Sadly, despite the hype, [`Next.js`](https://github.com/dwyl/learn-nextjs/issues/12) is not the ideal framework for building this. 
It makes building anything _interesting_ ([realtime](https://github.com/dwyl/phoenix-liveview-counter-tutorial)) way more complex than it needs to be. ⏳ 
What "interesting" features does a `Docs` site _need_...? 
**`Analytics`** without surrendering data/privacy to a 3<sup>rd</sup> party. 📈 
***Collaboration*** on editing docs. Without this feature we are stuck using tools that _do_ have realtime colab like `Miro`, `Figma`, `Notion`, `Google Docs`, etc. and then having to copy-paste into the docs repo. 🤦‍♂️ 

Anyway, if you _discover_ something please share. 🙏

**Comment by TheMikeyRoss** [comment]
oh don't get me started with next.js, I'm basically in a Stockholm Syndrome relationship with it. Once they deprecate the pages router then I'm out.

After that I'll probably look through vue/nuxt, Unlike Next.js, Nuxt seems to have a healthy progression pace and they're adding new experimental features left and right in production. Also, I think they also have a cool & calm community similar to Astro's

**Comment by zigang93** [comment]
Can try https://docusaurus.io/ from meta open source

**Comment by nelsonic** [comment]
@zigang93 we were using Docusaurus before Nextra. 🦕
 There are pros and cons to both. But neither fits our use case and Docusaurus is _even_ more convoluted to extend with the [features](https://github.com/dwyl/nextra-demo/issues/4#issuecomment-2295419071) we need.
Neither of these Docs tools has a “backend” so adding anything minimally advanced/custom is a chore where you have to “fight the framework”. 😔

If I had infinite time I would just _build_ something better. Alas time is scarce. ⏳😕

**Comment by nelsonic** [comment]
Yeah @TheMikeyRoss [I get the feeling that] very few people doing full-stack `JS` full-time 
actually _enjoy_ it or think it's the "right" tool for the job ... 
"_80% of programmers are NOT happy_": https://youtu.be/-pSf9_MgsZ4 **`Tech Debt`** ... 🧱 

> I say this as an [_instigator_](https://github.com/nelsonic/whistle) of `JS`-everywhere having used and championed `Node.js` since `v0.2.0` in `2010` ... ⏳ 🤯 
I _used_ to think that building _everything_ with `JS` was the [answer](https://www.laws-of-software.com/laws/atwood/) to the tech skills gap/shortage ... 👩‍💻 
but instead of _alleviating_ the problem it's gotten _much_ worse as the `JS` ecosystem has [fragmented](https://deno.com/blog/the-deno-company) 💔 

`Next.js` is a classic case of `Abilene paradox` https://en.wikipedia.org/wiki/Abilene_paradox 🔗 
where _nobody_ on the team thinks it's the "right" thing to do but no one speaks up about it. 🤐 

Anyway, I shouldn't "throw stones" ... If it works for you. Great! 🎉 
Meanwhile [`AI`](https://github.com/nelsonic/nelsonic.github.io/issues/733) will replace most programming jobs so none of this will matter. 🤖

**Comment by nelsonic** [comment]
Just tried searching for something (which I consider basic) in the `Nextra` docs ...
Decided to search `SO`: https://stackoverflow.com/questions/tagged/nextra
<img width="1398" alt="image" src="https://github.com/user-attachments/assets/94c984b0-5a37-4f80-b2ac-4c61e61c441f">
Really not a lot of questions/answers for a framework that has `10k+` stars. ⭐ 💭

**Comment by nelsonic** [comment]
One of the things that has annoyed me a _lot_ about `Nextra` is having _multiple_ `_meta.json` files (one per directory) to _configure_ how the link text appears in the Side and Top Nav: 
e.g: 
![image](https://github.com/user-attachments/assets/ed54f8ae-2299-4f56-80f9-cc54108f7703)

I would _much_ rather have a _single_ `sitemap.yml` (or `sitemap.json`) file at the root of the project that allows _optionally_ configuring all pages regardless of nesting level and avoiding duplication of files and effort. 
I often have 3+ `_meta.json` files open simultaneously and it's a context-switching pain. 🙃

**Comment by nelsonic** [comment]
One of the things that has annoyed me a _lot_ about `Nextra` is having _multiple_ `_meta.json` files (one per directory) to _configure_ how the link text appears in the Side and Top Nav: 
e.g: 

![nextra-multiple-meta-files](https://github.com/user-attachments/assets/000090b4-5f2c-4406-b73f-e4b769f16d9e)

I would _much_ rather have a _single_ `sitemap.yml` (or `sitemap.json`) file at the root of the project that allows _optionally_ configuring all pages regardless of nesting level and avoiding duplication of files and effort. 
I often have 3+ `_meta.json` files open simultaneously and it's a context-switching pain. 🙃

**Comment by nelsonic** [comment]
Same goes for having _many_ `/assets` directories spread across the project ... would definitely prefer a _single_ place to store all assets even if then organised into subdirs e.g: `/assets/engineering`, `/assets/client`, `/assets/frontend`, etc.

**Comment by nelsonic** [comment]
Yeah, I've noticed that I can easily reference a file in a root `/assets` dir but it's a hangover from `Docusaurus` ... 🙃 
The `_meta.json` per directory is lame and it might get _even_ more complex with allowing `JS` execution in the file: https://github.com/shuding/nextra/issues/852 ultimately, this could/should be streamlined.

**Comment by nelsonic** [comment]
These types of errors which give exactly zero insight as to _where_ the error occurred in the codebase ... 🤦‍♂️ 

<img width="1189" alt="image" src="https://github.com/user-attachments/assets/1837ee28-8943-404f-9fec-1b2adb5fb8c6">

```
Unhandled Runtime Error

Error: Hydration failed because the initial UI does not match what was rendered on the server.
See more info here: https://nextjs.org/docs/messages/react-hydration-error
```

What is a _non-technical_ person attempting to edit a document in `Nextra` meant to do with this? 🤷‍♂️ 
Why does a documentation site even _need_ to have _any_ client-side rendering? 😕

**Comment by nelsonic** [comment]
Another markup error crashing the whole site with no _helpful_ stack trace: 
<img width="1161" alt="image" src="https://github.com/user-attachments/assets/71ae4238-7bd5-402c-a2a5-37e9b6461760">

The page in question `/pages/index.mdx` has only one `<span>` tag and it is appropriately closed:
<img width="980" alt="image" src="https://github.com/user-attachments/assets/6f91ae36-4474-4dea-973a-1407088369be">

I've scoured this page for several minutes and not found any tag that does not have the appropriate _closing_ tag.
And yet, still get this silly error. 🤦‍♂️

**Comment by lacymorrow** [comment]
You need more newlines. MDX is injecting a <p> tag somewhere.

**Comment by lacymorrow** [comment]
Nextra 3 was announced...

https://the-guild.dev/blog/nextra-3
